client/src/connection_cache.rs

```diff
@@ -1,475 +1,117 @@
-        streamer::StakedNodes,
-        tls_certificates::new_self_signed_tls_certificate,
-    },
-    solana_tpu_client::{
-        connection_cache_stats::{ConnectionCacheStats, CONNECTION_STAT_SUBMISSION_INTERVAL},
-        tpu_connection_cache::MAX_CONNECTIONS,
+    solana_connection_cache::{
+        client_connection::ClientConnection as BlockingClientConnection,
+        connection_cache::{ConnectionCache as BaseConnectionCache, NewConnectionConfig},
+        nonblocking::client_connection::ClientConnection as NonblockingClientConnection,
     },
+    solana_quic_client::{QuicConfig, QuicConnectionManager},
+    solana_sdk::{pubkey::Pubkey, signature::Keypair},
+    solana_streamer::streamer::StakedNodes,
+    solana_udp_client::UdpConnectionManager,
     std::{
-        error::Error,
-        net::{IpAddr, Ipv4Addr, SocketAddr, UdpSocket},
-        sync::{atomic::Ordering, Arc, RwLock},
+        net::{IpAddr, Ipv4Addr, SocketAddr},
+        sync::{Arc, RwLock},
     },
 };
 
+pub const DEFAULT_CONNECTION_POOL_SIZE: usize = 4;
+pub const DEFAULT_CONNECTION_CACHE_USE_QUIC: bool = true;
+pub const MAX_CONNECTIONS: usize = 1024;
+
+/// A thin wrapper over connection-cache/ConnectionCache to ease
+/// construction of the ConnectionCache for code dealing both with udp and quic.
+/// For the scenario only with udp or quic, use connection-cache/ConnectionCache directly.
 pub struct ConnectionCache {
-    map: RwLock<IndexMap<SocketAddr, ConnectionPool>>,
-    stats: Arc<ConnectionCacheStats>,
-    last_stats: AtomicInterval,
-    connection_pool_size: usize,
-    tpu_udp_socket: Arc<UdpSocket>,
-    client_certificate: Arc<QuicClientCertificate>,
     use_quic: bool,
-    maybe_staked_nodes: Option<Arc<RwLock<StakedNodes>>>,
-    maybe_client_pubkey: Option<Pubkey>,
-
-    // The optional specified endpoint for the quic based client connections
-    // If not specified, the connection cache we create as needed.
-    client_endpoint: Option<Endpoint>,
-}
-
-/// Models the pool of connections
-struct ConnectionPool {
-    /// The connections in the pool
-    connections: Vec<Arc<BaseTpuConnection>>,
-
-    /// Connections in this pool share the same endpoint
-    endpoint: Option<Arc<QuicLazyInitializedEndpoint>>,
-}
-
-impl ConnectionPool {
-    /// Get a connection from the pool. It must have at least one connection in the pool.
-    /// This randomly picks a connection in the pool.
-    fn borrow_connection(&self) -> Arc<BaseTpuConnection> {
-        let mut rng = thread_rng();
-        let n = rng.gen_range(0, self.connections.len());
-        self.connections[n].clone()
-    }
-
-    /// Check if we need to create a new connection. If the count of the connections
-    /// is smaller than the pool size.
-    fn need_new_connection(&self, required_pool_size: usize) -> bool {
-        self.connections.len() < required_pool_size
-    }
+    cache: BaseConnectionCache,
```

Comment:
Why can't we just have BaseConnectionCache return its type (quic vs udp); seems kind of redundant to also store is_quic as a bool.

Reply:
BaseConnectionCache is supposed to be agnostic of either quic or udp. Secondly, is_quic is an input parameter here as convenient way to control the construction of either a quic or udp based connection cache.

Reply:
Apologies, meant in connection_manager since it's already a dyn. It can tell the connection cache backend and well as the connection cache wrapper whether it's supposed to use UDP or Quic as needed.

Reply:
My original thought was the ConnectionManager trait is also intended to be generic. Defining is_quic or enum of connection types will break the encapsulation. i.e introducing a new protocol will have to make modification in the generic ConnectionManager layer, imagining one day we decided tcp is better protocol for certain workflow. The generic ConnectionCache should work if all declared traits are implemented without any change.

Closely examining the refactored client/ConnectionCache code also shows its main job is to construct ConnectionCache of various protocol types based on user's input or the default settings, so it innately already knows what is the protocol type it is using -- and I thought it was okay to store that as a member for that fact.

Having said that, I am thinking, if the client are directly using the connection-cache/ConnectionCache directly -- do they need to know what the protocol type at the runtime? Maybe like how that function is used... I will define a enum of ProtocolType in connection-cache and force the implementor to return the type. The only downside is when a new protocol is introduced, we need to extend that enum. Which is okay, it will be very rare to add new protocol implementation.

Great point!